code/modules/lighting/light_effect_cast.dm

```diff
@@ -0,0 +1,303 @@
+	// but if it wont be rotated then we just check if its negative to flip (like the x flip)
+	var/x_flip
+	var/y_flip
+	if(xy_swap)
+		x_flip = y_offset > 0 ? -1 : 1
+		y_flip = x_offset < 0 ? -1 : 1
+	else
+		x_flip = x_offset < 0 ? -1 : 1
+		y_flip = y_offset < 0 ? -1 : 1
+
+	M.Scale(x_flip, y_flip)
+
+	//here we do the actual rotate if needed
+	if(xy_swap)
+		M.Turn(90)
+
+	//warning: you are approaching shitcode (this is where we move the shadow to the correct quadrant based on its rotation and flipping)
+	//shadows are only as big as a quarter or half of the light for optimization
+
+	//please for the love of god change this if there's a better way
+
+	if(num == 1)
+		if((x_flip == 1 && y_flip == 1 && xy_swap == 0) || (x_flip == -1 && y_flip == 1 && xy_swap == 1))
+			M.Translate(shadowoffset, shadowoffset)
+		else if((x_flip == 1 && y_flip == -1 && xy_swap == 0) || (x_flip == 1 && y_flip == 1 && xy_swap == 1))
+			M.Translate(shadowoffset, 0)
+		else if((xy_swap == 0 && x_flip == -y_flip) || (xy_swap == 1 && x_flip == -1 && y_flip == -1))
+			M.Translate(0, shadowoffset)
+	else
+		if(x_flip == 1 && y_flip == 1 && xy_swap == 0)
+			M.Translate(0, shadowoffset)
+		else if(x_flip == 1 && y_flip == 1 && xy_swap == 1)
+			M.Translate(shadowoffset / 2, shadowoffset / 2)
+		else if(x_flip == 1 && y_flip == -1 && xy_swap == 1)
+			M.Translate(-shadowoffset / 2, shadowoffset / 2)
+
+	//apply the transform matrix
+	I.transform = M
+	I.layer = LIGHTING_LAYER
+	//and add it to the lights overlays
+	temp_appearance += I
+
+// No direct TILE_BOUND shadow.
+/atom/movable/light/shadow/cast_main_shadow(turf/target_turf, x_offset, y_offset)
+	return
+
+/atom/movable/light/proc/cast_turf_shadow(var/turf/target_turf, var/x_offset, var/y_offset)
+	var/targ_dir = get_dir(target_turf, src)
+	// CHECK: may not actually smoothout that well.
+	var/blocking_dirs = 0
+	for(var/d in cardinal)
+		var/turf/T = get_step(target_turf, d)
+		if(CheckOcclusion(T))
+			blocking_dirs |= d
+
+	// The "edge" of the light, with images consisting of directional sprites from wall_lighting.dmi "pushed" in the correct direction.
+	var/image/I = image('icons/lighting/wall_lighting.dmi', loc = get_turf(src))
+	I.icon_state = "[blocking_dirs]-[targ_dir]"
+	I.pixel_x = (world.icon_size * light_range) + (x_offset * world.icon_size)
+	I.pixel_y = (world.icon_size * light_range) + (y_offset * world.icon_size)
+	I.layer = ABOVE_LIGHTING_LAYER
+	temp_appearance += I
+
+/atom/movable/light/proc/update_appearance()
+	overlays = temp_appearance
+	temp_appearance = null
+
+// On how many turfs do we cast a shadow ?
+/atom/movable/light/proc/cast_shadows()
+	//no shadows
+	if(light_range < 2 || light_type == LIGHT_DIRECTIONAL)
+		return
+
+	var/list/visible_turfs = list()
+
+	for(var/turf/T in view(light_range, src))
+		visible_turfs += T
+
+	for(var/turf/T in visible_turfs)
+		if(CheckOcclusion(T))
+			CastShadow(T)
```

Comment:
Is there a reason you use `visible_turfs` here instead of just looping through `view()` directly in the second loop?